pandas/core/groupby/ops.py

```diff
@@ -657,7 +657,15 @@ def _aggregate_series_pure_python(self, obj: Series, func):
             res = func(group)
             if result is None:
                 if isinstance(res, (Series, Index, np.ndarray)):
-                    raise ValueError("Function does not reduce")
+                    if len(res) == 1:
+                        # e.g. test_agg_lambda_with_timezone lambda e: e.head(1)
+                        # FIXME: are we potentially losing import res.index info?
+
+                        # TODO: use `.item()` if/when we un-deprecate it.
+                        # For non-Series we could just do `res[0]`
+                        res = next(iter(res))
```

Comment:
Why is this actually needed? For any later iteration, the len-1 `res` is just assigned below with `result[label] = res`, so why does it need to be unpacked for the first group?

Reply:
We might be able to get rid of this, but at this stage the goal is just to make the behavior match libreduction._extract_result

Reply:
But in the cython version that calls `_extract_result`, this is done for each group, not just the first (so in that sense it still doesn't match that)

Reply:
>  But in the cython version that calls _extract_result, this is done for each group, not just the first (so in that sense it still doesn't match that)

This is correct.  The more closely matching behavior is that only the first group is checked for array-like (there's also a discrepancy in what types of arraylikes are checked)  (there's  also^2 a discrepancy in that Reducer.get_result does a `res = res.values` check that is similar to _extract_result but not quite the same)